arch/arm/core/cortex_m/irq_manage.c

```diff
@@ -132,33 +132,17 @@ void z_irq_spurious(const void *unused)
 #ifdef CONFIG_PM
 void _arch_isr_direct_pm(void)
 {
-#if defined(CONFIG_ARMV6_M_ARMV8_M_BASELINE)
 	unsigned int key;
 
-	/* irq_lock() does what we want for this CPU */
+	/* Disable all interrupts except ZLIs. */
```

Comment:
`_arch_isr_direct_pm()`  could be called in ZLI and non-ZLI direct ISRs. A race will lead to `pm_system_resume()` not being called when it should be or corrupt the data accessed inside the function call, right?

On ARMv6, `irq_lock()` is nothing but disable all interrupts. On ARMv7, `cpsid i` needs to be used to disable all interrupts

Zephyr Controller calls  `ISR_DIRECT_PM()` here: https://github.com/zephyrproject-rtos/zephyr/blob/aaa7e7238576854745fe8ffd924bd7ae052b5a8b/subsys/bluetooth/controller/ll_sw/nordic/lll/lll.c#L106


Does this PR mean, no ZLI ISR call `ISR_DIRECT_PM()` ?

Reply:
Yes, ZLIs must not call any zephyr APIs since it breaks synchronisation :) pm_system_resume() is a zephyr API

Reply:
To be pedantic: Zephyr APIs are not guaranteed to work in a ZLI, and specifically as you point out anything that relies on synchronization is guaranteed not to work (or not to work reliably).  But there's lots of good stuff in the Zephyr API that you can use as long as you know what you're doing.  Like, Best Practice for communication with a ZLI almost certainly involves the Zephyr atomics API.  We've just been too lazy/conservative to try to document where the boundary is.

Reply:
> Zephyr Controller calls ISR_DIRECT_PM() here:

Right, but who runs the Zephyr controller with `CONFIG_PM` enabled? none of the platforms where the controller runs use `CONFIG_PM` I think, so that's probably why you've never seen an issue.

Reply:
This however is worrisome IMHO:
https://github.com/zephyrproject-rtos/zephyr/blob/aaa7e7238576854745fe8ffd924bd7ae052b5a8b/subsys/bluetooth/controller/ll_sw/nordic/lll/lll.c#L114

If you return 1 there, the kernel scheduling code will execute, which should never be done from a ZLI.

Reply:
@nordic-krch agreed! Could you follow-up with #17301 so I can add the assert after?

Reply:
if that is an issue we can add `ZLI_ASSERT` that will do `while(1)` or reset. The thing is that it will be fail immediately.

Reply:
Well, if it's just a loop you could simply disable interrupts and spin. This should work for all applications in principle? Then perhaps it should be overridable.